extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/NotFoundExceptionMapper.java

```diff
@@ -145,50 +157,58 @@ public Response toResponse(NotFoundException exception) {
+        }
+
+        return Response.status(Status.NOT_FOUND).build();
     }
 
     private Response respond(List<ResourceDescription> descriptions) {
-        if (headers.getAcceptableMediaTypes().contains(MediaType.APPLICATION_JSON_TYPE)) {
-            ErrorMessage errorMessage = new ErrorMessage();
-            errorMessage.errorMessage = "Resource Not Found";
-            errorMessage.existingResourcesDetails = descriptions;
-            return Response.status(Status.NOT_FOUND).entity(errorMessage).type(MediaType.APPLICATION_JSON_TYPE).build();
+        Variant variant = selectVariant(headers);
+
+        if (variant == JSON_VARIANT) {
+            return Response.status(Status.NOT_FOUND).type(MediaType.APPLICATION_JSON).build();
         }
 
-        TemplateHtmlBuilder sb = new TemplateHtmlBuilder();
-        sb.header("Resource Not Found", "REST interface overview");
-        for (ResourceDescription resource : descriptions) {
-            sb.resourcePath(resource.basePath);
-            for (MethodDescription method : resource.calls) {
-                sb.method(method.method, method.fullPath);
-                if (method.consumes != null) {
-                    sb.consumes(method.consumes);
-                }
-                if (method.produces != null) {
-                    sb.produces(method.produces);
+        if (variant == HTML_VARIANT) {
+            TemplateHtmlBuilder sb = new TemplateHtmlBuilder("404 - Resource Not Found", "", "REST interface overview");
+            sb.resourcesStart();
+            for (ResourceDescription resource : descriptions) {
+                sb.resourcePath(resource.basePath);
+                for (MethodDescription method : resource.calls) {
+                    sb.method(method.method, method.fullPath);
+                    if (method.consumes != null) {
+                        sb.consumes(method.consumes);
+                    }
+                    if (method.produces != null) {
+                        sb.produces(method.produces);
+                    }
+                    sb.methodEnd();
                 }
-                sb.methodEnd();
+                sb.resourceEnd();
             }
-            sb.resourceEnd();
-        }
-        if (descriptions.isEmpty()) {
-            sb.noResourcesFound();
+            if (descriptions.isEmpty()) {
+                sb.noResourcesFound();
+            }
+            sb.resourcesEnd();
+
+            return Response.status(Status.NOT_FOUND).entity(sb.toString()).type(MediaType.TEXT_HTML_TYPE).build();
         }
 
-        return Response.status(Status.NOT_FOUND).entity(sb.toString()).build();
+        return Response.status(Status.NOT_FOUND).build();
     }
 
-    public static class ErrorMessage {
-        public String errorMessage;
-        public List<ResourceDescription> existingResourcesDetails;
+    private static Variant selectVariant(HttpHeaders headers) {
+        ServerDrivenNegotiation negotiation = new ServerDrivenNegotiation();
```

Comment:
Why did you use this instead of the regular jax-rs method? Did it give you different results?

Reply:
TBH, I did that while trying to understand what was going wrong and then I thought it was better to keep it this way as the JAX-RS requests do more things than strictly necessary.

If this class is not stable enough, we can change it later to the version you had.

Reply:
No, I just wanted to know if it behaved differently. I expect the JAX-RS idiomatic way probably uses this underneath.

Reply:
It behaves exactly the same, it uses this class underneath but it does a bit more things. I trimmed it down to what we really wanted.